### DISCUSSION TITLE
[Proposal] Extensions for member of a class

### DISCUSSION THREAD

**Discussion by KeyserDSoze** [dotnet/csharplang#5223] [Language Ideas]
I have a class, for instance

```
public class MyClass
{
    public string MyString { get; init; }
    public MyProperty MyProperty { get; init; }
}
```

with MyProperty class, for instance
```
public class MyProperty
{
    public string Value { get; init; }
}
```
or as a record
```
public record MyProperty(string Value);
```
I'd like to create an extension like:
```
public static class MyClassExtensions
{
    public static string GetValue(this MyProperty myProperty)
        where MyProperty in MyClass
        => myProperty.Value;
}
```
to allow using it in this way
```
public static class Sample
{
    public static void Run()
    {
        var myClass = new MyClass { MyString = string.Empty, MyProperty = new MyProperty() { Value = string.Empty } };
        //I'd see the extension method only here
        myClass.MyProperty.GetValue();

        var myProperty = new MyProperty { Value = string.Empty };
        //and I wouldn't see here
        myProperty.GetValue();
    }
}
```
It's very useful, in my opinion, to have the chance to restrict the area of effect of extension methods.
I know it's possible to do it through scoping the class.
```
public static class MyClassExtensions
{
    public static string GetValueOfMyProperty(this MyClass myClass)
        => myClass.MyProperty.Value;
}
```
But I want to avoid it and this feature could be used in another way
```
public class MyClass
{
    public string MyString { get; init; }
    public MyProperty MyProperty { get; init; }
}
public class MyClass2
{
    public string MyString { get; init; }
    public MyProperty MyProperty { get; init; }
}
public class MyClass3
{
    public string MyString { get; init; }
    public MyProperty MyProperty { get; init; }
}
```
I could avoid the use on, for instance, MyClass2 in this way, in one shot
```
public static class MyClassExtensions
{
    public static string GetValue(this MyProperty myProperty)
        where MyProperty in MyClass, MyClass3
        => myProperty.Value;
}
```
avoiding a forced duplication like the next one
```
public static class MyClassExtensions
{
    public static string GetValueOfMyProperty(this MyClass myClass)
        => myClass.MyProperty.Value;
    public static string GetValueOfMyProperty(this MyClass3 myClass)
        => myClass.MyProperty.Value;
}
```

**Comment by PathogenDavid** [comment]
This sounds a lot like shapes https://github.com/dotnet/csharplang/discussions/164

**Comment by svick** [comment]
> It's very useful, in my opinion, to have the chance to restrict the area of effect of extension methods.

Can you explain why do you find it useful? Maybe a concrete example would help.

**Reply by KeyserDSoze** [reply]
I happened, several times, the need to give a different behaviour through extension method. 
On second thought, what I want to bring to the attention is the need of a scope or context for extension methods.
For instance, in my next example, I would give to final user of my library the chance to use my extension method for transform rapidly an object to a json string only when it is declared out of a container object.
```
public static class Runner
{
    public static void Run()
    {
        MyObject myObject = new();
        myObject.ToJson();

        MySuperiorObject mySuperiorObject = new() { MyObject = new MyObject() };
        //I want to avoid this
        mySuperiorObject.MyObject.ToJson();
    }
}
public class MySuperiorObject
{
    public MyObject MyObject { get; set; }
}
public class MyObject
{
    public string A { get; set; }
    public string B { get; set; }
}
public static class JsonExtensions
{
    public static string ToJson<T>(this T myObject, JsonSerializerOptions options = default)
        where T : class
        where T is not in object
    {
        return JsonSerializer.Serialize(myObject, options);
    }
}
```

**Reply by 333fred** [reply]
Rather than shapes, this sounds like roles. https://github.com/dotnet/csharplang/issues/1711.

**Reply by svick** [reply]
@KeyserDSoze Are you saying that you want `mySuperiorObject.MyObject.ToJson();` to be invalid, but `var myObject = mySuperiorObject.MyObject; myObject.ToJson();` would still be valid? To me, that sounds very counterintuitive.

**Reply by KeyserDSoze** [reply]
I don't think it's counterintuitive, obviously, if I want to do that I'd have to use the decorator pattern.
```
public static class Runner
{
    public static void Run()
    {
        MyDecoratedObject myObject = new();
        myObject.ToJson();

        MySuperiorObject mySuperiorObject = new() { MyObject = new MyObject() };
        //it doesn't exist more
        mySuperiorObject.MyObject.ToJson();
    }
}
public class MySuperiorObject
{
    public MyObject MyObject { get; set; }
}
public class MyObject
{
    public string A { get; set; }
    public string B { get; set; }
}
public class MyDecoratedObject : MyObject
{
    public string ToJson(JsonSerializerOptions options = default)
    {
        return JsonSerializer.Serialize(this, options);
    }
}
```
I'd like to avoid the use of decorator. In a few cases it's still not possible, for instance if MyObject was a sealed class and it was in an external library.
Roles is near to this solution and probably it could resolve it, but I think the correct name of what I mean is Scope, in this particular situation.
I used to use decorator to change behaviour or improve behaviour of a class in a scope, and sometimes this chance to decorate something is simply not possible (as I said for a sealed for example). 
Extension methods are great to do this job well but you lose the scope and everything of that kind of object results decorated.
